src/helpers/cards.ts

```diff
@@ -321,6 +323,165 @@ export const changeCardStatus = async (
   return cardData.card;
 };
 
+/**
+ * Triggers a series of webhooks calls (CARD_TOKEN_LIFECYCLE) simulating the provisioning token creation
+ * lifecycle. When done, it saves the final token in the card information of the person.
+ *
+ * @param personId {string}
+ * @param cardId {string}
+ */
+export const createProvisioningToken = async (
+  personId: string,
+  cardId: string,
+): Promise<ProvisioningTokenStatusChangePayload> => {
+
+  if (!personId) {
+    throw new Error("You have to provide personId");
+  }
+  if (!cardId) {
+    throw new Error("You have to provide cardId");
+  }
+
+  const person = await db.getPerson(personId);
+  const cardData = person.account.cards.find(({ card }) => card.id === cardId);
+  if (!cardData) {
+    throw new Error("Card not found");
+  }
+
+  const { provisioningToken } = cardData;
+  let walletId;
+
+  // Deactivate existing token
+  if (provisioningToken) {
+    walletId = provisioningToken.client_wallet_account_id
+    const data = {
+      card_id: cardData.card.id,
+      token_status: ProvisioningTokenStatus.DEACTIVATED,
+      token_reference_id: provisioningToken.token_reference_id,
+      client_wallet_account_id: walletId,
+      wallet_type: "GOOGLE",
+      event_type: "TOKEN_STATUS_UPDATED",
+      message_reason: "TOKEN_DEACTIVATED",
+    };
+    await triggerWebhook(CardWebhookEvent.CARD_TOKEN_LIFECYCLE, data);
+  }
+
+  /**
+   * Now go through all creation lifecycle:
+   * - TOKEN_CREATED (INACTIVE)
+   * - LUK_REPLENISHMENT (INACTIVE)
+   * - DEVICE_PROVISIONING_RESULT (INACTIVE)
+   * - OTP_VERIFICATION_RESULT (ACTIVE)
+   */
+  let baseData: ProvisioningTokenStatusChangePayload = {
+    card_id: cardData.card.id,
+    token_reference_id: uuid.v4(),
+    client_wallet_account_id: walletId || uuid.v4(),
+    wallet_type: "GOOGLE",
+  };
+
+  // Token Creation
+  let payload: ProvisioningTokenStatusChangePayload = {
+    ...baseData,
+    token_status: ProvisioningTokenStatus.INACTIVE,
+    event_type: "TOKEN_CREATED",
+    message_reason: "TOKEN_CREATED",
+  }
+  await triggerWebhook(CardWebhookEvent.CARD_TOKEN_LIFECYCLE, payload);
```

Comment:
could these all `triggerWebhook`calls be better organised in a loop?

You could define some kind of array: 

```
const params = [
  {    
      token_status: ProvisioningTokenStatus.INACTIVE,
      message_reason: "LUK_REPLENISHMENT",
  {
      token_status: ProvisioningTokenStatus.INACTIVE,
      message_reason: "DEVICE_PROVISIONING_RESULT",
  }
  // and so on
];
```

and then just loop it:

```
for (const opt of params) {
    await triggerWebhook(CardWebhookEvent.CARD_TOKEN_LIFECYCLE, {...baseData, ...opt});
}
```

Reply:
b0d94e4